common/membership/ringpop/hostinfo.go

```diff
@@ -25,19 +25,31 @@
 // Package ringpop provides a service-based membership monitor
 package ringpop
 
+import (
+	"fmt"
+
+	"github.com/dgryski/go-farm"
+	rpmembership "github.com/temporalio/ringpop-go/membership"
+
+	"go.temporal.io/server/common/membership"
+)
+
+// hostInfo represents a host in a ring.
 type hostInfo struct {
-	addr   string // ip:port
-	labels map[string]string
+	addr           string            // ip:port
+	labels         map[string]string // this map is shared with ringpop
+	labelsChecksum uint64            // cache this for quick comparison
 }
 
+var _ rpmembership.Member = (*hostInfo)(nil)
+var _ membership.HostInfo = (*hostInfo)(nil)
+
 // newHostInfo creates a new *hostInfo instance
 func newHostInfo(addr string, labels map[string]string) *hostInfo {
-	if labels == nil {
-		labels = make(map[string]string)
-	}
 	return &hostInfo{
-		addr:   addr,
-		labels: labels,
+		addr:           addr,
+		labels:         labels,
+		labelsChecksum: checksumLabels(labels),
```

Comment:
Can host labels be updated by any node that _isn't_ that host? If not, we could use a simple version counter here instead of checksumming the labels. When adding a new label we'd simply bump the version

Reply:
where do we get the previous version from? or do you mean the node that updates its draining state should also update a version number?

first, I don't think there's an interface to change more than one label atomically. (annoyingly, the underlying `memberlist` has an api for that but `NodeLabels` does not. we could add it in our fork though.)

second, it seems inefficient to pass around a version number with a gossip protocol just for the convenience of local comparisons. the checksum thing is pretty simple.

if you really don't like the checksum, I can just have it compare values for all known labels instead. there's only three of them.

Reply:
Nah, I was just curious. It may not be simpler but it popped into my head. I didn't grok that only the labels were propagated by SWIM; for some reason I had it in my head that we were propagating the checksum as well